lib/rspec/matchers.rb

```diff
@@ -698,7 +698,11 @@ def all(expected)
     #       matchers (due to how the custom matcher DSL was evaluated in 2.x,
     #       `match` could not be used there), but is no longer needed in 3.x.
     def match(expected)
-      BuiltIn::Match.new(expected)
+      if ::String === expected
+        BuiltIn::Captures.new(expected)
+      else
+        BuiltIn::Match.new(expected)
+      end
```

Comment:
@myronmarston adding with_captures to Match is the approach I took in #849. Two subclasses of Match are created (MatchWithCaptures, MatchWithNamedCaptures) and the appropriate one is instantiated/returned when with_captures is invoked. If it's never invoked, the matcher remains an instance of Match. The choice between MatchWithCaptures/MatchWithNamedCaptures is made based on whether the expected_captures are a Hash or hash-matcher.